java-squid/src/main/java/org/sonar/java/se/SymbolicValueRelation.java

```diff
@@ -34,73 +38,154 @@ protected SymbolicValueRelation(SymbolicValue v1, SymbolicValue v2) {
       }
+      usedRelations.add(relation);
     }
-    for (int i = 0; i < knownRelations.size(); i++) {
-      SymbolicValueRelation relation = knownRelations.get(i);
-      SymbolicValueRelation combined = underHypothesis(relation);
-      if (combined != null) {
-        final List<SymbolicValueRelation> relations = new ArrayList<>(knownRelations);
-        relations.remove(i);
-        Boolean result = combined.impliedBy(relations);
-        if (result != null) {
-          return result;
-        }
+    return impliedBy(transitiveReduction(knownRelations, usedRelations), usedRelations);
+  }
+
+  private Collection<SymbolicValueRelation> transitiveReduction(Collection<SymbolicValueRelation> knownRelations, Set<SymbolicValueRelation> usedRelations) {
+    boolean changed = false;
+    List<SymbolicValueRelation> result = new ArrayList<>();
+    for (SymbolicValueRelation relation : knownRelations) {
+      boolean used = false;
+      if (v1.equals(relation.v1)) {
+        used = result.addAll(relation.combinedRelations(knownRelations, usedRelations));
+      } else if (v1.equals(relation.v2)) {
+        used = result.addAll(relation.symmetric().combinedRelations(knownRelations, usedRelations));
+      }
+      if (used) {
+        changed = true;
+      } else {
+        result.add(relation);
       }
     }
-    return null;
+    return changed ? result : Collections.<SymbolicValueRelation>emptyList();
   }
 
-  protected boolean hasSameOperand(SymbolicValueRelation relation) {
-    return (v1.equals(relation.v1) && v2.equals(relation.v2))
-      || (v1.equals(relation.v2) && v2.equals(relation.v1));
+  private Collection<SymbolicValueRelation> combinedRelations(Collection<SymbolicValueRelation> relations, Set<SymbolicValueRelation> usedRelations) {
+    List<SymbolicValueRelation> result = new ArrayList<>();
+    for (SymbolicValueRelation relation : relations) {
+      SymbolicValueRelation combined = null;
+      if (v2.equals(relation.v1) && !v1.equals(relation.v2)) {
+        combined = relation.combinedAfter(this);
+      } else if (v2.equals(relation.v2) && !v1.equals(relation.v2)) {
+        combined = relation.symmetric().combinedAfter(this);
+      }
+      if (combined != null && !usedRelations.contains(combined)) {
+        result.add(combined);
+      }
+    }
+    return result;
   }
 
   @Override
   public String toString() {
     StringBuilder buffer = new StringBuilder();
-    buffer.append('(');
     buffer.append(v1);
     buffer.append(' ');
     buffer.append(getOperand());
     buffer.append(' ');
     buffer.append(v2);
-    buffer.append(')');
     return buffer.toString();
   }
 
   protected abstract String getOperand();
 
+  /**
+   * @return a new relation, the negation of the receiver
+   */
   protected abstract SymbolicValueRelation inverse();
 
   /**
+   * @return a new relation, equivalent to the receiver with inverted parameters
+   */
+  protected abstract SymbolicValueRelation symmetric();
```

Comment:
Handling the issue of symmetricity should be done in another ticket. It is useful to resolve SONARJAVA-1456, but it should not be part of the same commit!